x-pack/legacy/plugins/siem/public/pages/case/components/user_action_tree/index.tsx

```diff
@@ -48,35 +71,220 @@ const UserAction = styled(EuiFlexGroup)`
       border-bottom: ${theme.eui.euiBorderThin};
       border-radius: ${theme.eui.euiBorderRadius} ${theme.eui.euiBorderRadius} 0 0;
     }
-    .userAction__content {
-      padding: ${theme.eui.euiSizeM} ${theme.eui.euiSizeL};
-    }
     .euiText--small * {
       margin-bottom: 0;
     }
   `}
 `;
+const MySpinner = styled(EuiLoadingSpinner)`
+  .euiLoadingSpinner {
+    margin-top: 1px; // yes it matters!
+  }
+`;
+
+const ContentWrapper = styled.div`
+  ${({ theme }) => css`
+    padding: ${theme.eui.euiSizeM} ${theme.eui.euiSizeL};
+  `}
+`;
 
-const renderUserActions = (userActions: UserActionItem[]) => {
-  return userActions.map(({ avatarName, children, title }, key) => (
-    <UserAction key={key} gutterSize={'none'}>
-      <EuiFlexItem grow={false}>
-        <EuiAvatar className="userAction__circle" name={avatarName} />
-      </EuiFlexItem>
-      <EuiFlexItem>
-        <EuiPanel className="userAction__panel" paddingSize="none">
-          <EuiText size="s" className="userAction__title">
-            {title}
-          </EuiText>
-          {children && <div className="userAction__content">{children}</div>}
-        </EuiPanel>
-      </EuiFlexItem>
-    </UserAction>
-  ));
-};
+export const UserActionTree = React.memo(
+  ({ data, onUpdateField, isLoadingDescription }: UserActionTreeProps) => {
+    const [
+      { data: comments, isLoading: isLoadingComment, isLoadingCommentId },
+      dispatchUpdateComment,
+    ] = useUpdateComment(data.comments);
+    const [commentUpdate, setCommentUpdate] = useState('');
+    const [description, setDescription] = useState(data.description);
+    const [editCommentId, setEditCommentId] = useState('');
+    const [isEditDescription, setIsEditDescription] = useState(false);
+    const renderButtons = useCallback(({ cancelAction, saveAction }) => {
+      return (
+        <EuiFlexGroup gutterSize="s" alignItems="center">
+          <EuiFlexItem grow={false}>
+            <EuiButtonEmpty size="s" onClick={cancelAction} iconType="cross">
+              {i18n.CANCEL}
+            </EuiButtonEmpty>
+          </EuiFlexItem>
+          <EuiFlexItem grow={false}>
+            <EuiButton color="secondary" fill iconType="save" onClick={saveAction} size="s">
+              {i18n.SAVE}
+            </EuiButton>
+          </EuiFlexItem>
+        </EuiFlexGroup>
+      );
+    }, []);
+    const renderUserActions = useMemo(() => {
+      const userActions: UserActionItem[] = comments.reduce(
```

Comment:
I think this `reduce` function is too big, but good news here, you already divided into components. So we can go a step further and create these components like UserActionAvatar, UserActionTitle and UserActionChildren. And maybe we can transform the data comment before so we can only do one map over the comments to build our action tree.